src/models/user

```diff
@@ -0,0 +1,218 @@
+
+const rate = mongoose.Schema({
+    userId: {
+        type: mongoose.Schema.Types.ObjectId,
+        ref: "User",
+        required: true
+    },
+    rate: {
+        type: number,
+        enum: [1, 2, 3, 4, 5],
+        required: true
+    }
+})
+
+const baseUser = mongoose.Schema({
+    email: {
+        type: String,
+        unique: true,
+        required: true
+    },
+    hashedPassword: {
+        type: String,
+        required: true
+    },
+    provider: {
+        type: String,
+        enum: ['email', 'google', 'twitter'],
+        default: 'email',
+        required: true
+    },
+    providerId: {
+        type: String,
+    },
+    isVerified: {
+        type: Boolean,
+        required: true
+    },
+    createdEvents: [{
+        type: mongoose.Schema.Types.ObjectId,
+        ref: "Event",
+
+    }],
+    followers: [{
+        type: mongoose.Schema.Types.ObjectId,
+        ref: "User",
+    }],
+    createdFunds: [{
+        type: mongoose.Schema.Types.ObjectId,
+        ref: "Fund"
+    }]
+
+}, { toJSON: { virtuals: true } }, { toObject: { virtuals: true } });
+
+const BaseUser = mongoose.model("User", baseUser);
+
+
+const user = mongoose.Schema({
+
+    firstName: {
+        type: string,
+        required: true
+    },
+    lastName: {
+        type: string,
+        required: true
+    },
+    // fullname: virtual property
+
+    profileImage: {
+        type: string,
+    },
+    dateOfBirth: {
+        type: string,
+        required: true
+    },
+    preferredCities: [{
+        type: string,
+        enum: cities
+    }],
+
+    interests: [{
+        type: string,
+        enum: listOfCategories,
+        required: true
+    }],
+    gender: {
+        type: string,
+        enum: ['male', 'female', 'other']
+    },
+    followedEvents: [{
+        type: mongoose.Schema.Types.ObjectId,
+        ref: "Event",
+
+    }],
+    followedFunds: [{
+        type: mongoose.Schema.Types.ObjectId,
+        ref: "Fund",
+    }],
+    followedUsers: [{
+        type: mongoose.Schema.Types.ObjectId,
+        ref: "User",
+    }],
+    followedOrganizations: [{
+        type: mongoose.Schema.Types.ObjectId,
+        ref: "Organization",
+
+    }]
+
+}]
+
+
+
+// virtual
+user.virtual('fullName').get(function () {
+    return this.firstName + ' ' + this.lastName;
+});
+
+const User = BaseUser.discriminator("User", user);
+
+
+
+
+const organization = mongoose.Schema({
+    name: {
+        type: string,
+        required: true
+    },
+    description: {
+        type: string,
+        required: true
+    },
+    coverImage: {
+        type: string,
+        required: true
+
+    },
+    categories: [{
+        type: string,
+        enum: listOfCategories
+    }],
+    city: {
+        type: string,
+        required: true
+
+    },
+    rates: [rate],
+    // rate : virtual property
+    websiteUrl: {
+        type: string,
+    }
+
+
+});
+
+// virtual
```

Comment:
lets remove this comment it does not add any meaning to the code following it. its clear that the following code create a virtual.